File: Leetcode 1301-1400/1340/1340.py
# Leetcode 1340: Jump Game V
# https://leetcode.com/problems/jump-game-v/
# Solved on 24th of May, 2026
class Solution:
    def maxJumps(self, arr: list[int], d: int) -> int:
        """
        Calculates the maximum number of indices that can be visited starting from any index.

        :param arr: List of integers representing the heights at each index.
        :param d: Maximum jump distance allowed in either direction.
        :return: The maximum number of indices that can be visited.
        """
        arrayLength = len(arr)
        memoArray = [0] * arrayLength

        def calculateJumps(currIndex):
            if memoArray[currIndex] != 0:
                return memoArray[currIndex]

            maxJumpsCount = 1

            for rightIndex in range(currIndex + 1, min(currIndex + d + 1, arrayLength)):
                if arr[rightIndex] >= arr[currIndex]:
                    break
                maxJumpsCount = max(maxJumpsCount, 1 + calculateJumps(rightIndex))

            for leftIndex in range(currIndex - 1, max(currIndex - d - 1, -1), -1):
                if arr[leftIndex] >= arr[currIndex]:
                    break
                maxJumpsCount = max(maxJumpsCount, 1 + calculateJumps(leftIndex))

            memoArray[currIndex] = maxJumpsCount
            return maxJumpsCount

        overallMaxJumps = 0
        for startIndex in range(arrayLength):
            overallMaxJumps = max(overallMaxJumps, calculateJumps(startIndex))

        return overallMaxJumps

File: Leetcode 1301-1400/1340/1340.java
// Leetcode 1340: Jump Game V
// https://leetcode.com/problems/jump-game-v/
// Solved on 24th of May, 2026
class Solution {
    /**
     * Calculates the maximum number of indices you can visit starting from any index.
     * 
     * @param arr The array of heights.
     * @param d The maximum jump distance.
     * @return The maximum number of indices that can be visited.
     */
    public int maxJumps(int[] arr, int d) {
        int n = arr.length;
        int[] dp = new int[n];
        int ans = 0;
        
        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, dfs(arr, d, i, dp));
        }
        
        return ans;
    }
    
    private int dfs(int[] arr, int d, int i, int[] dp) {
        if (dp[i] != 0) {
            return dp[i];
        }
        
        int max = 1;
        int n = arr.length;
        
        for (int j = i + 1; j <= Math.min(n - 1, i + d); j++) {
            if (arr[j] >= arr[i]) {
                break;
            }
            max = Math.max(max, 1 + dfs(arr, d, j, dp));
        }
        
        for (int j = i - 1; j >= Math.max(0, i - d); j--) {
            if (arr[j] >= arr[i]) {
                break;
            }
            max = Math.max(max, 1 + dfs(arr, d, j, dp));
        }
        
        dp[i] = max;
        return max;
    }
}